File: app/src/main/java/ru/art2000/calculator/currency_converter/CurrenciesAddFragment.java
package ru.art2000.calculator.currency_converter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Px;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import ru.art2000.calculator.R;
import ru.art2000.extensions.CurrencyItemWrapper;
import ru.art2000.extensions.OnListChangeListener;
import ru.art2000.helpers.CurrencyValuesHelper;

public class CurrenciesAddFragment extends Fragment {

    AddCurrenciesAdapter adapter;
    @Px
    int recyclerViewBottomPadding;
    private View v = null;
    private RecyclerView recycler;
    private TextView emptyView;
    private Context mContext;
    private EditCurrenciesActivity parent;

    void scrollToTop() {
        recycler.smoothScrollToPosition(0);
    }

    @SuppressLint("InflateParams")
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (v == null) {
            mContext = getActivity();
            parent = (EditCurrenciesActivity) getActivity();
            v = inflater.inflate(R.layout.modify_currencies_layout, null);
            emptyView = v.findViewById(R.id.empty_tv);
            recycler = v.findViewById(R.id.modify_currencies_list);
            recycler.setPadding(0, 0, 0, recyclerViewBottomPadding);
            LinearLayoutManager llm = new LinearLayoutManager(mContext);
            llm.setOrientation(RecyclerView.VERTICAL);
            recycler.setLayoutManager(llm);
            setNewList(CurrencyValuesHelper.hiddenList);
        }
        return v;
    }

    private ArrayList<CurrencyItemWrapper> searchByQuery(String query) {
        return CurrencyValuesHelper.findAllItems(mContext, CurrencyValuesHelper.hiddenList, query);
    }

    void filterList(String query) {
        adapter.filterData(query);
    }

    private void setNewList(ArrayList<CurrencyItemWrapper> list) {
        if (adapter != null) {
            adapter.setNewData(list);
        } else {
            adapter = new AddCurrenciesAdapter(list);
        }
        if (recycler.getAdapter() == null) {
            recycler.setAdapter(adapter);
        }
        toggleEmptyView();
    }

    private void toggleEmptyView() {
        if (adapter == null)
            return;
        if (adapter.getItemCount() == 0) {
            if (emptyView.getVisibility() == View.GONE) {
                emptyView.setText(getEmptyText());
                emptyView.setVisibility(View.VISIBLE);
            }
        } else {
            if (emptyView.getVisibility() == View.VISIBLE) {
                emptyView.setVisibility(View.GONE);
            }
        }
    }

    private String getEmptyText() {
        if (parent.barSearchView.getQuery().length() > 0 && CurrencyValuesHelper.hiddenList.size() > 0) {
            return getString(R.string.empty_text_no_currencies_found);
        } else {
            return getString(R.string.empty_text_all_currencies_added);
        }
    }

    void removeFromCurrentList(ArrayList<CurrencyItemWrapper> list) {
        adapter.currentList.removeAll(list);
    }

    private void updateCheckState(List<CurrencyItemWrapper> list) {
        for (int i = 0; i < list.size(); ++i) {
            RecyclerView.ViewHolder holder = recycler.findViewHolderForAdapterPosition(i);
            if (holder != null) {
                ((AddCurrenciesAdapter.Holder) holder).checkBox.setChecked(list.get(i).isSelected);
            }
        }
    }

    class AddCurrenciesAdapter extends RecyclerView.Adapter<AddCurrenciesAdapter.Holder> {

        List<CurrencyItemWrapper> currentList;
        String currentQuery = "";

        int size;
        int selectedCount;

        boolean shouldUpdateVisibility = true;

        AddCurrenciesAdapter(@NonNull ArrayList<CurrencyItemWrapper> list) {
            currentList = list;
            size = currentList.size();
            countSelectedItems();
            CurrencyValuesHelper.hiddenListChangeListener = new OnListChangeListener<CurrencyItemWrapper>() {
                @Override
                public void onItemsAdded(ArrayList<CurrencyItemWrapper> addedItems) {
                }

                @Override
                public void onItemsRemoved(ArrayList<CurrencyItemWrapper> removedItems) {
                    selectedCount = 0;
                }
            };
        }

        @NonNull
        @Override
        @SuppressLint("InflateParams")
        public AddCurrenciesAdapter.Holder onCreateViewHolder(@NonNull ViewGroup parent,
                                                              int viewType) {
            View item = LayoutInflater
                    .from(mContext)
                    .inflate(R.layout.item_add_currencies_list, null);
            return new Holder(item);
        }

        @Override
        public void onViewAttachedToWindow(@NonNull Holder holder) {
            int pos = holder.getBindingAdapterPosition();
            holder.checkBox.setChecked(currentList.get(pos).isSelected);
        }

        @Override
        public void onBindViewHolder(@NonNull AddCurrenciesAdapter.Holder holder, int position) {
            CurrencyItemWrapper currencyItem = currentList.get(position);
            holder.code.setText(currencyItem.code);
            holder.name.setText(currencyItem.nameResourceId);

            holder.checkBox.setOnCheckedChangeListener(null);
            holder.checkBox.setChecked(currencyItem.isSelected);
            Log.d("itemBind", currencyItem.code + "||" + currencyItem.isSelected);
            holder.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChecked) {
                    if (!currencyItem.isSelected) {
                        currencyItem.isSelected = true;
                        ++selectedCount;
                    }
                } else {
                    if (currencyItem.isSelected) {
                        currencyItem.isSelected = false;
                        --selectedCount;
                    }
                }
                if (shouldUpdateVisibility) {
                    parent.toggleElementsVisibility();
                }
            });

            holder.itemView.setOnClickListener(v ->
                    holder.checkBox.performClick());
        }

        @Override
        public int getItemCount() {
            return size;
        }

        ArrayList<CurrencyItemWrapper> getSelectedItems() {
            ArrayList<CurrencyItemWrapper> selectedItems = new ArrayList<>();
            for (CurrencyItemWrapper itemWrapper : currentList) {
                if (itemWrapper.isSelected) {
                    selectedItems.add(itemWrapper);
                }
            }
            return selectedItems;
        }

        private void countSelectedItems() {
            selectedCount = 0;
            for (CurrencyItemWrapper currencyItemWrapper : currentList) {
                if (currencyItemWrapper.isSelected) {
                    ++selectedCount;
                }
            }
        }

        void deselectAll() {
            for (CurrencyItemWrapper currencyItemWrapper : currentList) {
                currencyItemWrapper.isSelected = false;
            }
            shouldUpdateVisibility = false;
            updateCheckState(currentList);
            shouldUpdateVisibility = true;
            selectedCount = 0;
        }

        void selectAll() {
            for (CurrencyItemWrapper currencyItemWrapper : currentList) {
                currencyItemWrapper.isSelected = true;
            }
            shouldUpdateVisibility = false;
            updateCheckState(currentList);
            shouldUpdateVisibility = true;
            selectedCount = size;
        }

        boolean isSomethingSelected() {
            return selectedCount != 0;
        }

        boolean isAllSelected() {
            return selectedCount == size;
        }

        void setNewData() {
            currentList = CurrencyValuesHelper.hiddenList;
            size = currentList.size();
            countSelectedItems();
            notifyDataSetChanged();
            toggleEmptyView();
        }

        void setNewData(List<CurrencyItemWrapper> newList) {
            currentList = newList;
            size = currentList.size();
            countSelectedItems();
            notifyDataSetChanged();
            toggleEmptyView();
            parent.toggleElementsVisibility();
        }

        void reFilterData() {
            filterData(currentQuery);
        }

        void filterData(String filterString) {
            new Thread(() -> {
                ArrayList<CurrencyItemWrapper> list = searchByQuery(filterString);
                currentQuery = filterString;
                parent.runOnUiThread(() -> setNewData(list));
            }).start();
        }

        class Holder extends RecyclerView.ViewHolder {

            TextView code;
            TextView name;
            CheckBox checkBox;

            Holder(final View itemView) {
                super(itemView);
                code = itemView.findViewById(R.id.currency_code);
                checkBox = itemView.findViewById(R.id.checkbox_add);
                name = itemView.findViewById(R.id.currency_name);
            }

        }
    }

}

File: currency/src/androidMain/kotlin/ru/art2000/calculator/currency/view/CurrenciesAddFragment.kt
package ru.art2000.calculator.currency.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.core.util.Consumer
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.RecyclerView
import dev.androidbroadcast.vbpd.viewBinding
import ru.art2000.calculator.currency.R
import ru.art2000.calculator.currency.databinding.ItemAddCurrenciesListBinding
import ru.art2000.calculator.currency.databinding.ModifyCurrenciesLayoutBinding
import ru.art2000.calculator.currency.model.CurrencyItem
import ru.art2000.calculator.currency.model.getNameIdentifier
import ru.art2000.calculator.currency.vm.CurrenciesAddModel
import ru.art2000.calculator.currency.vm.CurrenciesSettingsModel
import ru.art2000.extensions.activities.IEdgeToEdgeFragment
import ru.art2000.extensions.collections.LiveList
import ru.art2000.extensions.collections.LiveList.LiveListObserver
import ru.art2000.extensions.collections.calculateDiff
import ru.art2000.extensions.fragments.CommonReplaceableFragment
import ru.art2000.extensions.views.addOrientationItemDecoration

internal class CurrenciesAddFragment :
    CommonReplaceableFragment(R.layout.modify_currencies_layout), IEdgeToEdgeFragment {

    private val binding by viewBinding(ModifyCurrenciesLayoutBinding::bind)
    private val model: CurrenciesAddModel by activityViewModels<CurrenciesSettingsModel>()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val addCurrenciesAdapter = AddCurrenciesAdapter()
        binding.modifyCurrenciesList.apply {
            layoutManager = ru.art2000.extensions.views.OrientationManger(requireContext()) {
                addCurrenciesAdapter.itemCount == 0
            }
            addOrientationItemDecoration()
            emptyViewGenerator = { ctx, _, _ ->
                ru.art2000.extensions.views.createTextEmptyView(ctx, emptyTextRes)
            }
            emptyViewHolderBinder = Consumer { view: View? ->
                val emptyView = view as TextView?
                emptyView?.setText(emptyTextRes)
            }

            adapter = addCurrenciesAdapter
        }
    }

    override val bottomViews: List<View>
        get() = listOf(binding.modifyCurrenciesList)

    override fun onReselected() {
        binding.modifyCurrenciesList.smoothScrollToPosition(0)
    }

    @get:StringRes
    private val emptyTextRes: Int
        get() = if (model.currentQuery.isNotEmpty() && model.hiddenItems.value.isNotEmpty()) {
            R.string.empty_text_no_currencies_found
        } else {
            R.string.empty_text_all_currencies_added
        }

    private inner class AddCurrenciesAdapter : RecyclerView.Adapter<AddCurrenciesAdapter.Holder>() {

        init {
            model.displayedHiddenItems.observe(
                viewLifecycleOwner,
                object : LiveListObserver<CurrencyItem>() {
                    override fun onAnyChanged(previousList: List<CurrencyItem>) {
                        dispatchListUpdate(previousList, model.displayedHiddenItems)
                    }
                })
            model.selectedHiddenItems.observe(
                viewLifecycleOwner,
                object : LiveListObserver<CurrencyItem>() {
                    override fun onItemsInserted(
                        previousList: List<CurrencyItem>,
                        liveList: LiveList<CurrencyItem>,
                        insertedIndices: List<Int>
                    ) {
                        super.onItemsInserted(previousList, liveList, insertedIndices)
                        for (i in insertedIndices) {
                            markItem(liveList[i], true)
                        }
                    }

                    override fun onItemsRemoved(
                        previousList: List<CurrencyItem>,
                        removedIndices: List<Int>
                    ) {
                        super.onItemsRemoved(previousList, removedIndices)
                        for (i in removedIndices) {
                            markItem(previousList[i], false)
                        }
                    }
                })
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
            val binding = ItemAddCurrenciesListBinding.inflate(
                LayoutInflater.from(requireContext()), parent, false
            )
            return Holder(binding)
        }

        override fun onBindViewHolder(holder: Holder, position: Int) {
            val currencyItem = model.displayedHiddenItems[position]
            holder.bind(currencyItem)
        }

        override fun getItemCount() = model.displayedHiddenItems.size

        private fun markItem(currencyItem: CurrencyItem, selected: Boolean) {
            val holder = binding.modifyCurrenciesList.findViewHolderForAdapterPosition(
                model.displayedHiddenItems.indexOf(currencyItem)
            ) as Holder? ?: return
            holder.checkBox.isChecked = selected
        }

        private fun dispatchListUpdate(oldData: List<CurrencyItem>, newData: List<CurrencyItem>) {
            val result = calculateDiff(oldData, newData,
                { code == it.code }, { position == it.position })
            result.dispatchUpdatesTo(this)
        }

        private inner class Holder(viewBinding: ItemAddCurrenciesListBinding) :
            RecyclerView.ViewHolder(viewBinding.root) {

            val code = viewBinding.currencyCode
            val name = viewBinding.currencyName
            val checkBox = viewBinding.checkboxAdd

            init {
                itemView.setOnClickListener { checkBox.performClick() }
                checkBox.setOnClickListener {
                    val pos = bindingAdapterPosition
                    val item = model.displayedHiddenItems[pos]
                    if (checkBox.isChecked != model.isHiddenItemSelected(item)) {
                        model.setHiddenItemSelected(item, checkBox.isChecked)
                    }
                }
            }

            fun bind(currencyItem: CurrencyItem) {
                code.text = currencyItem.code
                name.setText(currencyItem.getNameIdentifier(name.context))
                checkBox.isChecked = model.isHiddenItemSelected(currencyItem)
            }
        }
    }
}